File: internal/codexbridge/protocol.go
package codexbridge

import "encoding/json"

const TestedCodexVersion = "0.148.0"

const RequireStructuredHumanInput = "When progress requires an answer from the human, use the structured request_user_input tool."

type rpcEnvelope struct {
	JSONRPC string          `json:"jsonrpc,omitempty"`
	ID      json.RawMessage `json:"id,omitempty"`
	Method  string          `json:"method,omitempty"`
	Params  json.RawMessage `json:"params,omitempty"`
	Result  json.RawMessage `json:"result,omitempty"`
	Error   *RPCError       `json:"error,omitempty"`
}

type RPCError struct {
	Code    int             `json:"code"`
	Message string          `json:"message"`
	Data    json.RawMessage `json:"data,omitempty"`
}

func (e *RPCError) Error() string { return e.Message }

type ClientInfo struct {
	Name    string `json:"name"`
	Title   string `json:"title,omitempty"`
	Version string `json:"version"`
}

type InitializeCapabilities struct {
	ExperimentalAPI bool `json:"experimentalApi"`
}

type InitializeParams struct {
	ClientInfo   ClientInfo             `json:"clientInfo"`
	Capabilities InitializeCapabilities `json:"capabilities"`
}

type ThreadStartParams struct {
	CWD                   string `json:"cwd"`
	DeveloperInstructions string `json:"developerInstructions"`
}

type ThreadResumeParams struct {
	ThreadID string `json:"threadId"`
	CWD      string `json:"cwd,omitempty"`
}

type Thread struct {
	ID    string `json:"id"`
	CWD   string `json:"cwd,omitempty"`
	Turns []Turn `json:"turns,omitempty"`
}

type ThreadResponse struct {
	Thread Thread `json:"thread"`
}

type TextInput struct {
	Type string `json:"type"`
	Text string `json:"text"`
}

type TurnStartParams struct {
	ThreadID            string      `json:"threadId"`
	Input               []TextInput `json:"input"`
	ClientUserMessageID string      `json:"clientUserMessageId,omitempty"`
}

type Turn struct {
	ID     string       `json:"id"`
	Status string       `json:"status,omitempty"`
	Items  []ThreadItem `json:"items,omitempty"`
	Error  *TurnError   `json:"error,omitempty"`
}

type TurnError struct {
	Message           string          `json:"message"`
	CodexErrorInfo    json.RawMessage `json:"codexErrorInfo,omitempty"`
	AdditionalDetails string          `json:"additionalDetails,omitempty"`
}

type TurnResponse struct {
	Turn Turn `json:"turn"`
}

type ThreadItem struct {
	Type     string `json:"type"`
	ID       string `json:"id"`
	ClientID string `json:"clientId,omitempty"`
	Text     string `json:"text,omitempty"`
	Phase    string `json:"phase,omitempty"`
}

type ItemCompletedNotification struct {
	ThreadID string     `json:"threadId"`
	TurnID   string     `json:"turnId"`
	Item     ThreadItem `json:"item"`
}

type ThreadReadParams struct {
	ThreadID     string `json:"threadId"`
	IncludeTurns bool   `json:"includeTurns"`
}

type TurnSteerParams struct {
	ThreadID            string      `json:"threadId"`
	ExpectedTurnID      string      `json:"expectedTurnId"`
	Input               []TextInput `json:"input"`
	ClientUserMessageID string      `json:"clientUserMessageId,omitempty"`
}

type TurnSteerResponse struct {
	TurnID string `json:"turnId"`
}

type TurnNotification struct {
	ThreadID string `json:"threadId"`
	Turn     Turn   `json:"turn"`
}

type Notification struct {
	Method string
	Params json.RawMessage
}

type ServerRequest struct {
	ID     json.RawMessage
	Method string
	Params json.RawMessage
}

type RequestUserInputOption struct {
	Label       string `json:"label"`
	Description string `json:"description"`
}

type RequestUserInputQuestion struct {
	ID       string                   `json:"id"`
	Header   string                   `json:"header"`
	Question string                   `json:"question"`
	Options  []RequestUserInputOption `json:"options"`
	IsOther  bool                     `json:"isOther"`
	IsSecret bool                     `json:"isSecret"`
}

type RequestUserInputParams struct {
	ThreadID   string                     `json:"threadId"`
	TurnID     string                     `json:"turnId"`
	ItemID     string                     `json:"itemId"`
	IsBlocking bool                       `json:"isBlocking"`
	Questions  []RequestUserInputQuestion `json:"questions"`
}

type NetworkApprovalContext struct {
	Host     string `json:"host"`
	Protocol string `json:"protocol"`
}

type NetworkPolicyAmendment struct {
	Action string `json:"action"`
	Host   string `json:"host"`
}

type CommandApprovalParams struct {
	ThreadID                        string                  `json:"threadId"`
	TurnID                          string                  `json:"turnId"`
	ItemID                          string                  `json:"itemId"`
	ApprovalID                      string                  `json:"approvalId"`
	Command                         string                  `json:"command"`
	CWD                             string                  `json:"cwd"`
	Reason                          string                  `json:"reason"`
	CommandActions                  json.RawMessage         `json:"commandActions"`
	NetworkApprovalContext          *NetworkApprovalContext `json:"networkApprovalContext"`
	ProposedExecpolicyAmendment     []string                `json:"proposedExecpolicyAmendment"`
	ProposedNetworkPolicyAmendments []json.RawMessage       `json:"proposedNetworkPolicyAmendments"`
}

type FileChangeApprovalParams struct {
	ThreadID  string `json:"threadId"`
	TurnID    string `json:"turnId"`
	ItemID    string `json:"itemId"`
	Reason    string `json:"reason"`
	GrantRoot string `json:"grantRoot"`
}

type PermissionApprovalParams struct {
	ThreadID    string          `json:"threadId"`
	TurnID      string          `json:"turnId"`
	ItemID      string          `json:"itemId"`
	CWD         string          `json:"cwd"`
	Reason      string          `json:"reason"`
	Permissions json.RawMessage `json:"permissions"`
}

type MCPElicitationParams struct {
	ThreadID        string          `json:"threadId"`
	TurnID          string          `json:"turnId"`
	ServerName      string          `json:"serverName"`
	Mode            string          `json:"mode"`
	Message         string          `json:"message"`
	RequestedSchema json.RawMessage `json:"requestedSchema"`
	URL             string          `json:"url"`
	ElicitationID   string          `json:"elicitationId"`
}

File: crates/hq-codex/src/protocol.rs
//! Handwritten private DTO subset checked against the pinned generated schema.

use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct WireError {
    #[allow(dead_code)]
    pub code: i64,
    #[allow(dead_code)]
    pub message: String,
    #[serde(default)]
    #[allow(dead_code)]
    pub data: Value,
}

#[derive(Debug, Deserialize)]
pub(crate) struct WireMessage {
    #[serde(default)]
    pub id: Option<Value>,
    #[serde(default)]
    pub method: Option<String>,
    #[serde(default)]
    pub params: Option<Value>,
    #[serde(default)]
    pub result: Option<Value>,
    #[serde(default)]
    pub error: Option<WireError>,
}

#[derive(Serialize)]
pub(crate) struct ClientRequest<'a, T> {
    pub method: &'a str,
    pub id: u64,
    pub params: T,
}

#[derive(Serialize)]
pub(crate) struct ClientNotification<'a, T> {
    pub method: &'a str,
    pub params: T,
}

#[derive(Serialize)]
pub(crate) struct ClientResult<T> {
    pub id: Value,
    pub result: T,
}

#[derive(Serialize)]
pub(crate) struct ClientError<'a> {
    pub id: Value,
    pub error: ClientErrorBody<'a>,
}

#[derive(Serialize)]
pub(crate) struct ClientErrorBody<'a> {
    pub code: i64,
    pub message: &'a str,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct InitializeParams<'a> {
    pub client_info: ClientInfo<'a>,
    pub capabilities: InitializeCapabilities,
}

#[derive(Serialize)]
pub(crate) struct ClientInfo<'a> {
    pub name: &'a str,
    pub title: &'a str,
    pub version: &'a str,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct InitializeCapabilities {
    pub experimental_api: bool,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct ThreadStartParams<'a> {
    pub cwd: &'a str,
    pub developer_instructions: &'a str,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub model: Option<&'a str>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub approval_policy: Option<&'a str>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sandbox: Option<&'a str>,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct ThreadResumeParams<'a> {
    pub thread_id: &'a str,
    pub cwd: &'a str,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub model: Option<&'a str>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub approval_policy: Option<&'a str>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sandbox: Option<&'a str>,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct ThreadReadParams<'a> {
    pub thread_id: &'a str,
    pub include_turns: bool,
}

#[derive(Debug, Deserialize)]
pub(crate) struct ThreadResponse {
    pub thread: Thread,
}

#[derive(Clone, Debug, Deserialize)]
pub(crate) struct Thread {
    pub id: String,
    #[serde(default)]
    pub turns: Vec<Turn>,
}

#[derive(Clone, Debug, Deserialize)]
pub(crate) struct Turn {
    pub id: String,
    #[serde(default, deserialize_with = "string_or_default")]
    pub status: String,
    #[serde(default)]
    pub items: Vec<ThreadItem>,
    #[serde(default)]
    #[allow(dead_code)]
    pub error: Option<TurnError>,
}

#[derive(Clone, Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct TurnError {
    #[serde(default, deserialize_with = "string_or_default")]
    #[allow(dead_code)]
    pub message: String,
}

#[derive(Clone, Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct ThreadItem {
    #[serde(rename = "type")]
    pub kind: String,
    pub id: String,
    #[serde(default, deserialize_with = "string_or_default")]
    pub client_id: String,
    #[serde(default, deserialize_with = "string_or_default")]
    pub text: String,
    #[serde(default, deserialize_with = "string_or_default")]
    pub phase: String,
    #[serde(default, deserialize_with = "string_or_default")]
    pub status: String,
    #[serde(default, deserialize_with = "string_or_default")]
    pub command: String,
    #[serde(default)]
    pub aggregated_output: Option<String>,
    #[serde(default)]
    pub exit_code: Option<i64>,
    #[serde(default)]
    pub changes: Vec<FileChange>,
    #[serde(default, deserialize_with = "string_or_default")]
    pub server: String,
    #[serde(default, deserialize_with = "string_or_default")]
    pub tool: String,
    #[serde(default, deserialize_with = "string_or_default")]
    pub query: String,
}

#[derive(Clone, Debug, Deserialize)]
pub(crate) struct FileChange {
    pub path: String,
    #[serde(default)]
    pub diff: String,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct TextInput<'a> {
    #[serde(rename = "type")]
    pub kind: &'static str,
    pub text: &'a str,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct TurnStartParams<'a> {
    pub thread_id: &'a str,
    pub input: [TextInput<'a>; 1],
    pub client_user_message_id: String,
}

#[derive(Debug, Deserialize)]
pub(crate) struct TurnResponse {
    pub turn: Turn,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct TurnSteerParams<'a> {
    pub thread_id: &'a str,
    pub expected_turn_id: &'a str,
    pub input: [TextInput<'a>; 1],
    pub client_user_message_id: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct TurnSteerResponse {
    pub turn_id: String,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct TurnInterruptParams<'a> {
    pub thread_id: &'a str,
    pub turn_id: &'a str,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct TurnNotification {
    pub thread_id: String,
    pub turn: Turn,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct ItemNotification {
    pub thread_id: String,
    pub turn_id: String,
    pub item: ThreadItem,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct PlanNotification {
    pub thread_id: String,
    pub turn_id: String,
    #[serde(default)]
    pub explanation: Option<String>,
    #[serde(default)]
    pub plan: Vec<PlanStep>,
}

#[derive(Debug, Deserialize)]
pub(crate) struct PlanStep {
    pub step: String,
    pub status: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct DiffNotification {
    pub thread_id: String,
    pub turn_id: String,
    #[serde(default)]
    pub diff: Option<String>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct ProgressNotification {
    pub thread_id: String,
    pub turn_id: String,
    pub item_id: String,
    #[serde(default, alias = "delta")]
    pub message: String,
}

fn string_or_default<'de, D>(deserializer: D) -> Result<String, D::Error>
where
    D: serde::Deserializer<'de>,
{
    Option::<String>::deserialize(deserializer).map(Option::unwrap_or_default)
}
